sherpa-onnx/csrc/offline-recognizer-funasr-nano-impl.cc

```diff
@@ -0,0 +1,492 @@
+  fake_token_len = audio_token_len;
+  int64_t pad_id = tokenizer_->GetPadTokenId();
+  if (pad_id < 0) pad_id = tokenizer_->GetEosTokenId();
+  std::vector<int64_t> source_ids;
+  source_ids.reserve(ids_before.size() + audio_token_len + ids_after.size());
+  source_ids.insert(source_ids.end(), ids_before.begin(), ids_before.end());
+  // Use pad tokens as placeholders for audio embeddings
+  source_ids.insert(source_ids.end(), audio_token_len, pad_id);
+  source_ids.insert(source_ids.end(), ids_after.begin(), ids_after.end());
+  return source_ids;
+}
+
+// Sample token from logits using greedy decoding (argmax).
+// Handles both FP16 and FP32 logits, skipping NaN/Inf values.
+// Returns token ID 0 as fallback if all logits are invalid.
+int64_t OfflineRecognizerFunASRNanoImpl::SampleTokenFromLogitsFp16OrFp32(
+    const void *logits, bool is_fp16, int32_t vocab_size) const {
+  int32_t best = 0;
+  float best_val = -1e30f;
+  bool found_valid = false;
+  if (is_fp16) {
+    const uint16_t *p = reinterpret_cast<const uint16_t *>(logits);
+    for (int32_t i = 0; i < vocab_size; ++i) {
+      float v = Fp16ToFp32(p[i]);
+      if (std::isfinite(v) && v > best_val) {
+        best_val = v;
+        best = i;
+        found_valid = true;
+      }
+    }
+  } else {
+    const float *p = reinterpret_cast<const float *>(logits);
+    for (int32_t i = 0; i < vocab_size; ++i) {
+      if (std::isfinite(p[i]) && p[i] > best_val) {
+        best_val = p[i];
+        best = i;
+        found_valid = true;
+      }
+    }
+  }
+  if (!found_valid) {
+    return 0;
+  }
+  return static_cast<int64_t>(best);
+}
+
+// Generate text from encoder output using autoregressive decoding with KV cache.
+// Combines text embeddings (from prompts) and audio embeddings (from encoder)
+// to form the input sequence, then generates tokens autoregressively.
+OfflineRecognitionResult OfflineRecognizerFunASRNanoImpl::GenerateText(
+    Ort::Value encoder_out, const std::string &system_prompt,
+    const std::string &user_prompt) const {
+  OfflineRecognitionResult result;
+  if (!model_->HasEmbeddingModel()) {
+    SHERPA_ONNX_LOGE("Embedding model is required but not provided.");
+    result.text = "";
+    return result;
+  }
+  auto memory_info =
+      Ort::MemoryInfo::CreateCpu(OrtDeviceAllocator, OrtMemTypeDefault);
+  const auto &funasr_config = config_.model_config.funasr_nano;
+  auto enc_shape = encoder_out.GetTensorTypeAndShapeInfo().GetShape();
+  int32_t audio_token_len = static_cast<int32_t>(enc_shape[1]);
+  int32_t hidden_size = static_cast<int32_t>(enc_shape[2]);
+  int32_t fbank_beg_idx = 0;
+  int32_t fake_token_len = 0;
+  std::vector<int64_t> source_ids = BuildSourceIds(
+      system_prompt, user_prompt, audio_token_len, fbank_beg_idx,
+      fake_token_len);
+  int32_t context_len = static_cast<int32_t>(source_ids.size());
+  const int32_t max_seq_len = 2048;
+  if (context_len > max_seq_len) {
+    source_ids.resize(max_seq_len);
```

Comment:
Will the recognition result be truncated as a result of this resize?
If so, could you print a log message noting that the result may be truncated due to excessive input length?